hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/interfaces/ContainerDeletionChoosingPolicyTemplate.java

```diff
@@ -0,0 +1,83 @@
+package org.apache.hadoop.ozone.container.common.interfaces;
+
+
+import com.google.common.base.Preconditions;
+import org.apache.hadoop.hdds.scm.container.common.helpers.StorageContainerException;
+import org.apache.hadoop.ozone.container.common.impl.ContainerData;
+import org.apache.hadoop.ozone.container.keyvalue.KeyValueContainerData;
+import org.apache.hadoop.ozone.container.keyvalue.statemachine.background.BlockDeletingService.ContainerBlockInfo;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Abstract class that serves as the template for deletion choosing policy.
+ */
+public abstract class ContainerDeletionChoosingPolicyTemplate
+    implements ContainerDeletionChoosingPolicy {
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(ContainerDeletionChoosingPolicyTemplate.class);
+
+  @Override
+  public final List<ContainerBlockInfo> chooseContainerForBlockDeletion(
+      int blockCount, Map<Long, ContainerData> candidateContainers)
+      throws StorageContainerException {
+    Preconditions.checkNotNull(candidateContainers,
+        "Internal assertion: candidate containers cannot be null");
+
+    int originalBlockCount = blockCount;
+    List<ContainerBlockInfo> result = new ArrayList<>();
+    List<KeyValueContainerData> orderedList = new LinkedList<>();
+
+    for (ContainerData entry: candidateContainers.values()) {
+      orderedList.add((KeyValueContainerData) entry);
+    }
+
+    orderByDescendingPriority(orderedList);
+
+    // Here we are returning containers based on blockCount which is basically
+    // number of blocks to be deleted in an interval. We are also considering
+    // the boundary case where the blocks of the last container exceeds the
+    // number of blocks to be deleted in an interval, there we return that
+    // container but with container we also return an integer so that total
+    // blocks don't exceed the number of blocks to be deleted in an interval.
+
+    for (KeyValueContainerData entry : orderedList) {
+      if (entry.getNumPendingDeletionBlocks() > 0) {
+        long numBlocksToDelete = Math.min(blockCount,
+            entry.getNumPendingDeletionBlocks());
+        blockCount -= numBlocksToDelete;
+        result.add(new ContainerBlockInfo(entry, numBlocksToDelete));
+        if (LOG.isDebugEnabled()) {
+          LOG.debug("Select container {} for block deletion, "
+              + "pending deletion blocks num: {}.", entry.getContainerID(),
+              entry.getNumPendingDeletionBlocks());
+        }
+        if (blockCount == 0) {
+          break;
+        }
+      } else {
+        LOG.debug("Stop looking for next container, there is no"
+            + " pending deletion block contained in remaining containers.");
+        break;
+      }
```

Comment:
This optimization is valid for TopN (which is sorted by number of pending deletion blocks), but not for Random.  As far as I see this early exit was not present previously in either implementation.  Let's omit this.

Reply:
Thank you for the review. Yes, you are right. I got it mixed up with some old version.